apollo-federation/src/sources/connect/validation/graphql/strings.rs

```diff
@@ -0,0 +1,158 @@
+//! Helpers for dealing with GraphQL literal strings and getting locations within them.
+//!
+//! Specifically, working around these issues with `apollo-compiler` which make determining the
+//! line/column of locations _within_ a string impossible:
+//!
+//! 1. Quotes are included in locations (i.e., `line_column_range`) but not the values from `Node<Value>`.
+//! 2. String values in `Node<Value>` are stripped of insignificant whitespace
+
+use std::ops::Range;
+
+use apollo_compiler::ast::Value;
+use apollo_compiler::parser::LineColumn;
+use apollo_compiler::parser::SourceMap;
+use apollo_compiler::Node;
+use line_col::LineColLookup;
+
+#[derive(Clone, Copy)]
+pub(crate) struct GraphQLString<'schema> {
+    /// The GraphQL String literal without quotes, but with all whitespace intact
+    raw_string: &'schema str,
+    /// Where `raw_string` _starts_ in the source text
+    offset: usize,
+}
+
+impl<'schema> GraphQLString<'schema> {
+    /// Get the raw string value of this GraphQL string literal
+    ///
+    /// Returns `None` if the value was not a string literal or if location data is messed up
+    pub fn new(value: &'schema Node<Value>, sources: &'schema SourceMap) -> Option<Self> {
+        let value_without_quotes = value.as_str()?;
+        let source_span = value.location()?;
+        let file = sources.get(&source_span.file_id())?;
+        let source_text = file.source_text();
+        let start_of_quotes = source_span.offset();
+        let end_of_quotes = source_span.end_offset();
+        let value_with_quotes = source_text.get(start_of_quotes..end_of_quotes)?;
+
+        // On each line, the whitespace gets messed up for multi-line strings
+        // So we find the first and last lines of the parsed value (no whitespace) within
+        // the raw value (with whitespace) to get our raw string.
+        let first_line_of_value = value_without_quotes.lines().next()?;
+        let start_of_value = value_with_quotes.find(first_line_of_value)?;
+        let last_line_of_value = value_without_quotes.lines().last()?;
+        let end_of_value = value_with_quotes.rfind(last_line_of_value)? + last_line_of_value.len();
+        let raw_string = value_with_quotes.get(start_of_value..end_of_value)?;
+
+        Some(Self {
+            raw_string,
+            offset: start_of_quotes + start_of_value,
+        })
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.raw_string
+    }
+
+    pub fn line_col_for_subslice(
+        &self,
+        substring_location: Range<usize>,
+        lookup: &LineColLookup,
+    ) -> Option<Range<LineColumn>> {
+        let start_offset = self.offset + substring_location.start;
+        let end_offset = self.offset + substring_location.end;
+
+        let (line, column) = lookup.get(start_offset);
```

Comment:
This call panics if the offset is incorrect, which is very worrisome. You need to either document this in a `# Panics` rustdoc on this method, or you could validate the offsets within this method with a fallible `Result`. But really this is making me question this dependency even more.

Reply:
If we lookup beyond the length of the input, that's a bug in our implementation. We could make a `Code::Internal`, maybe, which indicates as much?

Reply:
Right, it's just a matter of how we handle such a bug. An internal error code would be fine. If it's really a bug that should never happen, it could arguably be a panic, and be documented with `Panics` - there are a few other cases of that around the router, but there are also places where lints are ensuring no panics, so I'm not sure what the most consistent handling would be here.